File: app/src/main/java/com/example/atlimer/MainActivity.java
package com.example.atlimer;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.widget.Button;
import android.widget.Toast;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends Activity {

    private static final int BT_PERMISSION_REQUEST = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // ⚠️ PERMISSION ELŐSZÖR
        if (!hasBluetoothPermission()) {
            requestBluetoothPermission();
            return;
        }

        // CSAK EZUTÁN jöhet UI / BT
        setContentView(R.layout.activity_main);

        Button btnContinue = findViewById(R.id.btnContinue);

        btnContinue.setOnClickListener(v -> {
            startActivity(new Intent(this, DashboardActivity.class));
            finish();
        });
    }

    private boolean hasBluetoothPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true;

        return ContextCompat.checkSelfPermission(this,
                Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(this,
                Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestBluetoothPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ActivityCompat.requestPermissions(this,
                    new String[]{
                            Manifest.permission.BLUETOOTH_CONNECT,
                            Manifest.permission.BLUETOOTH_SCAN
                    },
                    BT_PERMISSION_REQUEST);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String[] permissions,
                                           int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == BT_PERMISSION_REQUEST) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                recreate(); // újraindítjuk az activity-t
            } else {
                Toast.makeText(this,
                        "Bluetooth permission required",
                        Toast.LENGTH_LONG).show();
                finish();
            }
        }
    }
}

File: app/src/main/java/com/example/atlimer/MainActivity.kt

package com.example.atlimer

import android.bluetooth.*
import android.content.*
import android.os.*
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var adapter: BluetoothAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        adapter = BluetoothAdapter.getDefaultAdapter()

        val scanView = findViewById<LinearLayout>(R.id.scanView)
        val dashView = findViewById<LinearLayout>(R.id.dashboardView)

        val refreshBtn = findViewById<Button>(R.id.refreshBtn)
        val listView = findViewById<ListView>(R.id.deviceList)

        val powerBtn = findViewById<Button>(R.id.powerBtn)
        val lightBtn = findViewById<Button>(R.id.lightBtn)
        val backBtn = findViewById<Button>(R.id.backBtn)

        fun showScan() {
            dashView.visibility = View.GONE
            scanView.visibility = View.VISIBLE
        }

        fun showDash() {
            scanView.visibility = View.GONE
            dashView.visibility = View.VISIBLE
        }

        fun loadDevices() {
            val devices = adapter.bondedDevices.toList()
            val names = devices.map { "${it.name} (${it.address})" }
            listView.adapter =
                ArrayAdapter(this, android.R.layout.simple_list_item_1, names)

            listView.setOnItemClickListener { _, _, pos, _ ->
                val dev = devices[pos]
                val intent = Intent(this, BtService::class.java)
                intent.putExtra("address", dev.address)
                startForegroundService(intent)
                showDash()
            }
        }

        refreshBtn.setOnClickListener { loadDevices() }

        powerBtn.setOnClickListener {
            sendBroadcast(Intent("POWER_TOGGLE"))
        }

        lightBtn.setOnClickListener {
            sendBroadcast(Intent("LIGHT_TOGGLE"))
        }

        backBtn.setOnClickListener {
            stopService(Intent(this, BtService::class.java))
            showScan()
        }

        loadDevices()
    }
}
